File: pkg/dbtable/dbms.py
from typing import Callable, List, Dict, Any, Callable
from collections import defaultdict


class Table:
    '''Database table'''
    def __init__(self, columns: List[str], types: List[type]) -> None:
        assert len(columns) == len(types)
        self.columns = columns # names of the columns
        self.types = types     # types of the columns
        self.rows = []         # List[Row] where Row = Dict[str, Any]

    def __repr__(self):
        if len(self) > 20:
            rows = '\n '.join(str(row) for row in self.rows[:20]) + '\n ...'
        else:
            rows = '\n '.join(str(row) for row in self.rows)
        return f'Table(\n {self.columns}\n {rows}\n)'

    def __getitem__(self, idx: int):
        '''Returns a Row'''
        return self.rows[idx]

    def __iter__(self):
        return iter(self.rows)

    def __len__(self):
        return len(self.rows)

    def col2type(self, col: str) -> type:
        '''helper method to get the type of a column'''
        return self.types[self.columns.index(col)]

    def insert(self, values: list) -> None:
        # basic checks before insertion
        if len(values) != len(self.types):
            raise ValueError(f'Please provide {len(self.types)} values')
        for value, typ in zip(values, self.types):
            if not isinstance(value, typ):
                raise TypeError(f'Expected type {typ} but got {value}')
        self.rows.append(dict(zip(self.columns, values)))

    def update(self, updates: Dict[str, Any], predicate: Callable = lambda row: True):
        # basic checks before updating
        for col, val in updates.items():
            if col not in self.columns:
                raise TypeError(f'Invalid column: {col}')
            typ = self.col2type(col)
            if not isinstance(val, typ) and val is not None:
                raise TypeError(f'Expected type {typ} but got {val}')
        for row in self.rows:
            if predicate(row):
                for col, val in updates.items():
                    row[col] = val

    def delete(self, predicate: Callable = lambda row: True):
        purgees = [row for row in self.rows if predicate(row)]
        for purgee in purgees:
            self.rows.remove(purgee)

    def select(self, 
        keep_cols: List[str] = None, 
        additional_cols: Dict[str, Callable] = None
    ):
        if keep_cols is None:
            keep_cols = self.columns
        if additional_cols is None:
            additional_cols = {}
        new_cols = keep_cols + list(additional_cols.keys())
        keep_types = [self.col2type(col) for col in keep_cols]
        add_types = [calc.__annotations__['return'] for calc in additional_cols.values()]
        # create a new table for the results
        tb = Table(new_cols, keep_types + add_types)
        for row in self.rows:
            new_row = [row[col] for col in keep_cols]
            for calc in additional_cols.values():
                new_row.append(calc(row))
            tb.insert(new_row)
        return tb

    def where(self, predicate: Callable = lambda row: True):
        ''''Return only those rows that satisfy the predicate'''
        tb = Table(self.columns, self.types)
        for row in self.rows:
            if predicate(row):
                values = [row[col] for col in self.columns]
                tb.insert(values)
        return tb

    def limit(self, nrows: int):
        ''''Return only the top nRows rows'''
        tb = Table(self.columns, self.types)
        for i, row in enumerate(self.rows, 1):
            if i > nrows:
                break
            values = [row[col] for col in self.columns]
            tb.insert(values)
        return tb

    def groupby(self,
        groupby_cols: List[str],
        aggregates: Dict[str, Callable],
        having: Callable = lambda group: True
    ):
        grouped_rows = defaultdict(list)
        for row in self.rows:
            key = tuple(row[col] for col in groupby_cols)
            grouped_rows[key].append(row)
        newcols = groupby_cols + list(aggregates.keys())
        groupby_types = [self.col2type(col) for col in groupby_cols]
        aggregate_types = [
            agg.__annotations__['return'] for agg in aggregates.values()
        ]
        tb = Table(newcols, groupby_types + aggregate_types)
        for key, rows in grouped_rows.items():
            if having(rows):
                newrow = list(key)
                for agg_name, agg_fn in aggregates.items():
                    newrow.append(agg_fn(rows))
                tb.insert(newrow)
        return tb

    def orderby(self, order: Callable) -> 'Table':
        tb = self.select()  # copy
        tb.rows.sort(key=order)
        return tb

    def join(self, other_table: 'Table', left_join=False) -> 'Table':
        # columns of both tables
        join_on_cols = [c for c in self.columns if c in other_table.columns]
        # columns only of right table
        additional_cols = [c for c in other_table.columns if c not in join_on_cols]
        # columns of left table and additional columns from right table
        newcols = self.columns + additional_cols
        newtypes = self.types + [other_table.col2type(c) for c in additional_cols]
        jtb = Table(newcols, newtypes)  # join table
        for row in self.rows:
            isjoin = lambda orow: all(orow[c] == row[c] for c in join_on_cols)
            other_rows = other_table.where(isjoin).rows
            # each other row that matches in the join-on columns produces a result row
            for orow in other_rows:
                jtb.insert(
                    [row[c] for c in self.columns] + [orow[c] for c in additional_cols]
                )
            if left_join and len(other_rows) == 0:
                jtb.insert(
                    [row[c] for c in self.columns] + ['' for _ in additional_cols]
                )
        return jtb

def min_userid(rows) -> int:
    return min(row['userID'] for row in rows)

def length(rows) -> int:
    return len(rows)

def first_letter(row) -> str:
    return row['name'][0] if row['name'] else ''

def avg_nfriends(rows) -> float:
    return sum(row['nfriends'] for row in rows) / len(rows)

def enough_friends(rows) -> bool:
    return avg_nfriends(rows) > 1

def sum_userids(rows) -> int:
    return sum(row['userID'] for row in rows)

def count_interests(rows) -> int:
    '''Counts the number of rows with entries in interests columns'''
    return len([row for row in rows if row['interest']])


if __name__ == '__main__':

    '''
    CREATE TABLE users (
        userID INT NOT NULL,
        name VARCHAR(200),
        nfriends INT    
    );
    '''
    users = Table(['userID', 'name', 'nfriends'], [int, str, int])
    data = [
        [0, "Hero", 0],
        [1, "Dunn", 2],
        [2, "Sue", 3],
        [3, "Chi", 3],
        [4, "Thor", 3],
        [5, "Clive", 2],
        [6, "Hicks", 3],
        [7, "Devin", 2],
        [8, "Kate", 2],
        [9, "Klein", 3],
        [10, "Jen", 1]
    ]
    for row in data:
        users.insert(row)
    print('created:\n', users)

    print('before update:', users[1])
    '''
    UPDATE users
    SET nfriends = 3
    WHERE userID = 1;
    '''
    users.update({'nfriends': 3}, predicate=lambda row: row['userID'] == 1)
    print('after update:', users[1])
    '''
    SELECT userID FROM users WHERE name = 'Dunn';
    '''
    dunn_ids = (
        users
        .where(lambda row: row['name'] == 'Dunn')
        .select(keep_cols=['userID'])
    )
    print('dunn ids:\n', dunn_ids)
    '''
    DELETE FROM users WHERE userID = 1;
    '''
    users.delete(lambda row: row['userID'] == 1)
    print('delete:\n', users)
    '''
    SELECT * FROM users;
    '''
    alldata = users.select()
    print('all data:\n', alldata)
    '''
    SELECT userID FROM users;
    '''
    ids = users.select(keep_cols=['userID', 'name'])
    print('select ids:\n', ids)
    '''
    SELECT LENGTH(name) AS nameLen FROM users;
    '''
    def name_len(row) -> int: return len(row['name'])
    lens = users.select(keep_cols=[], additional_cols={'nameLen': name_len})
    '''
    SELECT * FROM users LIMIT 4;
    '''
    four = users.limit(4)
    '''
    SELECT 
        LENGTH(name) AS name_len,
        MIN(userID) AS min_userID,
        COUNT(*) AS nUsers
    FROM users
    GROUP BY LENGTH(name);
    '''
    stats_by_len = (
        users
        .select(additional_cols={'name_len': name_len})
        .groupby(
            groupby_cols=['name_len'], 
            aggregates={'min_userid': min_userid, 'nusers': length} 
        )
    )
    print(stats_by_len)
    '''
    SELECT SUM(userID) AS userid
    FROM users
    WHERE userID > 1;
    '''
    userid_sum = (
        users
        .where(lambda row: row['userID'] > 1)
        .groupby(groupby_cols=[], aggregates={'userid_sum': sum_userids})
    )
    print(userid_sum)
    '''
    SELECT 
        SUBSTR(name, 1, 1) AS first_letter,
        AVG(nfriends) AS avg_nfriends
    FROM users
    GROUP BY SUBSTR(name, 1, 1)
    HAVING AVG(nfriends) > 1;
    '''
    avg_by_letter = (
        users
        .select(additional_cols={'first_letter': first_letter})
        .groupby(
            groupby_cols=['first_letter'], 
            aggregates={'avg_nfriends': avg_nfriends},
            having=enough_friends
        )
    )
    ''''
    SELECT * FROM users
    ORDER BY name
    LIMIT 4;
    '''
    friendliest = (
        avg_by_letter
        .orderby(order=lambda row: -row['avg_nfriends'])
        .limit(4)
    )
    '''
    CREATE TABLE user_interests (
        userID INT NOT NULL,
        interest VARCHAR(100) NOT NULL 
    )
    '''
    user_interests = Table(['userID', 'interest'], [int, str])
    user_interests.insert([0, 'SQL'])
    user_interests.insert([0, 'NoSQL'])
    user_interests.insert([2, 'SQL'])
    user_interests.insert([2, 'MySQL'])
    '''
    SELECT * FROM users
    LEFT JOIN user_interests
    ON users.userID = user_interests.userID
    '''
    joined = users.join(user_interests, left_join=True)
    '''
    SELECT users.name
    FROM users
    JOIN user_interests
    ON users.userID = user_interests.userID
    WHERE user_interests.interest = 'SQL'
    '''
    sql_users = (
        users
        .join(user_interests)
        .where(lambda row: row['interest'] == 'SQL')
        .select(keep_cols=['name'])
    )
    '''
    SELECT users.userID, COUNT(user_interests.interest) AS ninterests
    FROM users
    LEFT JOIN user_interests
    ON users.userID = user_interests.userID
    '''
    user_interest_counts = (
        users
        .join(user_interests, left_join=True)
        .groupby(groupby_cols=['userID'], aggregates={'ninterests': count_interests})
    )
    ''''
    SELECT MIN(userID) AS min_userID 
    FROM (SELECT userID FROM user_interests WHERE interests = 'SQL') sql_interests;
    '''
    sql_ids = (
        user_interests
        .where(lambda row: row['interest'] == 'SQL')
        .select(keep_cols=['userID'])
    )
    ids = sql_ids.groupby(groupby_cols=[], aggregates={'minUserIds': min_userid})

File: cmd/dbms.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"sort"
)

type Columns map[string]string

func (cols Columns) Copy() Columns {
	newCols := make(Columns)
	for k, v := range cols {
		newCols[k] = v
	}
	return newCols
}

func (cols Columns) Update(oCols Columns) {
	for k, v := range oCols {
		cols[k] = v
	}
}

type Row map[string]interface{}

func (row Row) Copy() Row {
	rowCopy := make(Row)
	for k, v := range row {
		rowCopy[k] = v
	}
	return rowCopy
}

func (row Row) Update(orow Row) {
	for k, v := range orow {
		row[k] = v
	}
}

type Predicate func(Row) bool

type Table struct {
	Cols    Columns
	Rows    []Row
	sortCol string
}

func NewTable(cols Columns, rows []Row) *Table {
	if rows == nil {
		return &Table{Cols: cols, Rows: make([]Row, 0, 20)}
	}
	return &Table{Cols: cols, Rows: rows}
}

// implmentation of sort.Interface
func (tb *Table) Len() int {
	return len(tb.Rows)
}

func (tb *Table) Swap(i, j int) {
	tb.Rows[i], tb.Rows[j] = tb.Rows[j], tb.Rows[i]
}

func (tb *Table) Less(i, j int) bool {
	sortType, ok := tb.Cols[tb.sortCol]
	if !ok {
		panic("no column to sort by")
	}
	left, right := tb.Rows[i][tb.sortCol], tb.Rows[j][tb.sortCol]
	switch sortType {
	case "int":
		return left.(int) < right.(int)
	case "float64":
		return left.(float64) < right.(float64)
	case "string":
		return left.(string) < right.(string)
	default:
		return true
	}
}

func (tb *Table) String() string {
	txt := fmt.Sprintf("Table(\n   %v\n", tb.Cols)
	for _, row := range tb.Rows {
		txt += fmt.Sprintf("   %v\n", row)
	}
	return txt + ")"
}

func (tb *Table) typeCheck(row Row) {
	for key, value := range row {
		typename := fmt.Sprintf("%T", value)
		if tb.Cols[key] != typename {
			if typename == "float64" && tb.Cols[key] == "int" {
				// fmt.Println("type: ", typename)
				row[key] = int(value.(float64))
			} else if typename != "<nil>" {
				log.Fatalf("[TypeError] expected %v got %v", tb.Cols[key], typename)
			}
		}
	}
}

func (tb *Table) Get(i int) Row {
	return tb.Rows[i]
}

func (tb *Table) Insert(row Row) {
	if len(row) != len(tb.Cols) {
		panic("[InsertError] Row does not have correct number of entries")
	}
	tb.typeCheck(row)
	tb.Rows = append(tb.Rows, row)
}

func (tb *Table) Update(update Row, predicate Predicate) {
	if predicate == nil {
		predicate = func(row Row) bool { return true }
	}
	for _, row := range tb.Rows {
		if predicate(row) {
			for k, v := range update {
				row[k] = v
			}
		}
	}
}

func (tb *Table) Delete(predicate Predicate) {
	newRows := make([]Row, 0, tb.Len())
	for _, row := range tb.Rows {
		if !predicate(row) {
			newRows = append(newRows, row)
		}
	}
	tb.Rows = newRows
}

func (tb *Table) Where(predicate Predicate) *Table {
	ntb := NewTable(tb.Cols, nil)
	for _, row := range tb.Rows {
		if predicate(row) {
			ntb.Insert(row.Copy())
		}
	}
	return ntb
}

func (tb *Table) Limit(nRows int) *Table {
	ntb := NewTable(tb.Cols, make([]Row, nRows))
	for i, row := range tb.Rows {
		if i >= nRows {
			break
		}
		ntb.Rows[i] = row.Copy()
	}
	return ntb
}

type Calculation func(Row) float64

func (tb *Table) Select(
	keepCols []string, // columns to keep of the old ones
	addCols map[string]Calculation, // additional columns to compute
) *Table {
	var newCols = make(Columns)
	if keepCols == nil {
		keepCols = make([]string, 0, len(tb.Cols))
		for key, value := range tb.Cols {
			newCols[key] = value
			keepCols = append(keepCols, key)
		}
	} else {
		for _, key := range keepCols {
			newCols[key] = tb.Cols[key]
		}
	}
	for key := range addCols {
		newCols[key] = "float64"
	}
	newRows := make([]Row, tb.Len())
	for i, row := range tb.Rows {
		newRow := make(Row)
		for _, key := range keepCols {
			newRow[key] = row[key]
		}
		for key, fun := range addCols {
			newRow[key] = fun(row)
		}
		newRows[i] = newRow
	}
	return NewTable(newCols, newRows)
}

type Aggregators map[string]func([]Row) float64

func (tb *Table) GroupBy(
	groupByCols []string,
	aggregates Aggregators,
	having func([]Row) bool,
) *Table {
	if having == nil {
		having = func(rows []Row) bool { return true }
	}
	newCols := make(Columns)
	for _, key := range groupByCols {
		newCols[key] = tb.Cols[key]
	}
	for key, _ := range aggregates {
		newCols[key] = "float64"
	}
	groups := make(map[string][]Row)
	for _, row := range tb.Rows {
		txt := ""
		for _, key := range groupByCols {
			txt += fmt.Sprintf("%v", row[key])
		}
		groups[txt] = append(groups[txt], row)
	}
	newRows := make([]Row, 0, len(groups))
	for _, rows := range groups {
		if !having(rows) {
			continue
		}
		row := rows[0]
		newRow := make(Row)
		for _, key := range groupByCols {
			newRow[key] = row[key]
		}
		for key, aggregate := range aggregates {
			newRow[key] = aggregate(rows)
		}
		newRows = append(newRows, newRow)
	}
	return NewTable(newCols, newRows)
}

func (tb *Table) OrderBy(sortCol string) *Table {
	ntb := tb.Select(nil, nil)
	ntb.sortCol = sortCol
	sort.Sort(ntb)
	return ntb
}

func (tb *Table) Join(otherTb *Table, leftJoin bool) *Table {
	joinCols := make([]string, 0)
	addCols := make(Columns)
	for col, typ := range otherTb.Cols {
		if _, ok := tb.Cols[col]; !ok {
			addCols[col] = typ
		} else {
			joinCols = append(joinCols, col)
		}
	}
	newCols := tb.Cols.Copy()
	newCols.Update(addCols)
	ntb := NewTable(newCols, nil)
	for _, row := range tb.Rows {
		isJoin := func(orow Row) bool {
			for _, col := range joinCols {
				if row[col] != orow[col] {
					return false
				}
			}
			return true
		}
		otherRows := make([]Row, 0)
		for _, orow := range otherTb.Rows {
			if isJoin(orow) {
				otherRows = append(otherRows, orow)
			}
		}
		for _, orow := range otherRows {
			newRow := row.Copy()
			newRow.Update(orow)
			ntb.Insert(newRow)
		}
		if leftJoin && len(otherRows) == 0 {
			newRow := row.Copy()
			for k, _ := range addCols {
				newRow[k] = nil
			}
			ntb.Insert(newRow)
		}
	}
	return ntb
}

func sumUids(rows []Row) float64 {
	sum := 0.0
	for _, row := range rows {
		sum += float64(row["userID"].(int))
	}
	return sum
}

func main() {

	users := NewTable(
		Columns{"userID": "int", "name": "string", "nfriends": "int"}, // columns
		nil,
	)

	bytes, err := ioutil.ReadFile("data.json")
	if err != nil {
		panic(err)
	}
	data := make([]Row, 0, 20)
	err = json.Unmarshal(bytes, &data)
	for _, row := range data {
		users.Insert(row)
	}
	fmt.Println("created:\n", users)
	fmt.Println("before update", users.Get(1))
	users.Update(
		Row{"nfriends": 3},
		func(row Row) bool { return row["userID"].(int) == 1 })
	fmt.Println("after update", users.Get(1))

	dunn := users.Where(func(row Row) bool { return row["nfriends"].(int) == 3 })
	fmt.Println("dunn ids:\n", dunn)

	users.Delete(func(row Row) bool { return row["userID"].(int) == 1 })
	fmt.Println("delete:\n", users)

	uids := users.Select([]string{"userID", "name"}, nil)
	fmt.Println("select:\n", uids)
	lowIds := uids.Where(func(row Row) bool { return row["userID"].(int) < 6 })
	fmt.Println("where:\n", lowIds)

	alldata := users.Select(nil, nil)
	fmt.Println("alldata:\n", alldata)

	small := users.Limit(4)
	fmt.Println("limit:\n", small)

	g := users.GroupBy(nil, Aggregators{"sumUids": sumUids}, nil)
	fmt.Println("groupby:\n", g)

	g = users.GroupBy([]string{"nfriends"}, Aggregators{"sumUids": sumUids}, nil)
	fmt.Println("groupby nfriends:\n", g)

	sorted := users.OrderBy("nfriends")
	fmt.Println("sorted:\n", sorted)

	uinterests := NewTable(
		Columns{"userID": "int", "interest": "string"},
		[]Row{
			Row{"userID": 0, "interest": "SQL"}, Row{"userID": 0, "interest": "NoSQL"},
			Row{"userID": 2, "interest": "SQL"},
		},
	)
	fmt.Println("user interests before insertion\n", uinterests)
	uinterests.Insert(Row{"userID": 2, "interest": "MySQL"})
	fmt.Println("user interests after insertion\n", uinterests)

	jtb := users.Join(uinterests, true)
	fmt.Println("join table\n", jtb)

}
